Support type cast in group by

clemvnt opened the pull request:
Hello,

Cast type support in group by.

This PR needs https://github.com/OData/odata.net/pull/2879.

### Issues

Fixes https://github.com/OData/AspNetCoreOData/issues/1180

### Description

For the following group by : `groupby((MyAddress/Fully.Qualified.Namespace.HomeAddress/HomeNO))`

The response was an error :

```
System.NotSupportedException: The query specified in the URI is not valid. Binding OData QueryNode of kind 'SingleResourceCast' is not supported by 'AggregationBinder'.
```

Now, the response is :

```json
[
  {
    "myAddress": {
        "homeNO": "Value1"
    }
  },
  {
    "myAddress": {
        "homeNO": "Value2"
    }
  }
]
```

clemvnt commented:
@microsoft-github-policy-service agree

wachugamaina commented on src/Microsoft.AspNetCore.OData/Query/Expressions/TransformationBinderBase.cs:
```diff
@@ -172,5 +176,16 @@ private Expression CreateOpenPropertyAccessExpression(SingleValueOpenPropertyAcc
+            IEdmStructuredTypeReference structured = node.StructuredTypeReference;
+            Contract.Assert(structured != null, "NS casts can contain only structured types");
+
+            Type clrType = Model.GetClrType(structured);
```
Could we stick with using var for consistency ?

clemvnt replied:
This method is based on the following method : https://github.com/OData/AspNetCoreOData/blob/7e2e70718d06d3e20d68cc233f4d29377911235a/src/Microsoft.AspNetCore.OData/Query/Expressions/QueryBinder.cs#L692-L703

There doesn't seem to be any consistency between explicit types and `var` in the project.

But if someone of the organization feels the same way, I'll make the changes.

wachugamaina commented on src/Microsoft.AspNetCore.OData/Query/Expressions/TransformationBinderBase.cs:
```diff
@@ -172,5 +176,16 @@ private Expression CreateOpenPropertyAccessExpression(SingleValueOpenPropertyAcc
+
+            Type clrType = Model.GetClrType(structured);
+
+            Expression source = Bind(node.Source);
```
Could we stick with using var for consistency ?

clemvnt replied:
See previous comment.

wachugamaina commented on test/Microsoft.AspNetCore.OData.Tests/Query/Query/ApplyQueryOptionTest.cs:
```diff
@@ -1535,6 +1535,142 @@ public void ApplyTo_Returns_Correct_Queryable_ForFilter(string filter, int[] cus
+            {
+                List<Product> productList = new List<Product>();
+
+                Product p1 = new DerivedProduct
```
Could we stick with using var for consistency ?

clemvnt replied:
See previous comment.

wachugamaina commented on test/Microsoft.AspNetCore.OData.Tests/Query/Query/ApplyQueryOptionTest.cs:
```diff
@@ -1584,6 +1720,22 @@ public IQueryable<Customer> Get()
         }
     }
+
+    public class ProductsForTypeCastController : ODataController
```
Could we document these public classes and methods ?

clemvnt replied:
Here, we're in a test file, so the other classes in this file have no comments.

But if someone of the organization feels the same way, I'll make the changes.

ElizabethOkerio commented:
@clemvnt Will you continue working on this PR?

clemvnt commented:
Yes. I'm waiting for a new version of Microsoft.OData.Core with the https://github.com/OData/odata.net/pull/2879 fix before publishing the PR.

clemvnt commented:
I've changed the base branch to main. This PR requires the next version of Microsoft.OData.Core (8.0.2). Other than that, it's ready.

clemvnt commented:
This PR is ready.

WanjohiSammy commented:
@clemvnt Are you still working on this PR?

clemvnt commented:
I fixed the merge conflict. The PR is ready.

cc. @WanjohiSammy

gathogojr commented on test/Microsoft.AspNetCore.OData.Tests/Query/Query/ApplyQueryOptionTest.cs:
```diff
@@ -1534,6 +1534,142 @@ public void ApplyTo_Returns_Correct_Queryable_ForFilter(string filter, int[] cus
+        // Arrange
+        var model = new ODataModelBuilder()
+                        .Add_Products_EntityType()
+                        .Add_DerivedProducts_EntityType()
+                        .Add_DerivedProducts_EntityType()
```
```suggestion
                        .Add_DerivedProducts_EntityType()
```

gathogojr commented:
@clemvnt Please check out the failing tests

Also test for derived complex type (in particular) as per scenario:
```
class A
{
    public string Prop { get; set; }
}

class DerivedFromA : A
{
    public string PropInDerived { get; set; }
}

class Product
{
    // ...
    public A A { get; set; }
    // ...
}

// ...
modelBuilder.ComplexType<A>();
modelBuilder.ComplexType<DerivedFromA>();
```

Request URL
```
Products?groupby((A/Fully.Qualified.Namespace.DerivedFromA/PropInDerived))
```

clemvnt commented:
* Merged with main.
* Fixed tests.
* Added a new test case : 
```
Products?groupby((Category/Microsoft.AspNetCore.OData.Tests.Models.DerivedCategory/DerivedCategoryName))`)
```

gathogojr commented:
> * Merged with main.
> * Fixed tests.
> * Added a new test case :
> 
> ```
> Products?groupby((Category/Microsoft.AspNetCore.OData.Tests.Models.DerivedCategory/DerivedCategoryName))`)
> ```

The specific scenario I asked you to verify is one where the derived type is a complex type, not entity type. Here's an illustration of a scenario that seem to proceed unexpected results:
```csharp
// Data models
namespace Microsoft.AspNetCore.OData.E2E.Tests.DollarApply
{    public class MyCustomer
    {
        public int Id { get; set; }
        public MyAddress Address { get; set; }
    }

    public class MyAddress
    {
        public string Street { get; set; }
    }

    public class MyHomeAddress : MyAddress
    {
        public string HomeNo { get; set; }
        public MyCity City { get; set; }
    }

    public class MyCity
    {
        public string Name { get; set; }
    }
}


// Edm model
var builder = new ODataConventionModelBuilder();
            
builder.EntitySet<MyCustomer>("MyCustomers");
builder.ComplexType<MyAddress>();
builder.ComplexType<MyHomeAddress>();
builder.ComplexType<MyCity>();

var model = builder.GetEdmModel();


// Controller
public class MyCustomersController : ODataController
{
    private static readonly List<MyCustomer> myCustomers = new List<MyCustomer>
    {
        new MyCustomer
        {
            Id = 1,
            Address = new MyHomeAddress { Street = "High Street", HomeNo = "#1", City = new MyCity { Name = "Belfast" } }
        },
        new MyCustomer
        {
            Id = 2,
            Address = new MyHomeAddress { Street = "Moore Street", HomeNo = "#2", City = new MyCity { Name = "Dublin" } }
        }
    };

    [EnableQuery]
    public ActionResult<IEnumerable<MyCustomer>> Get()
    {
        return myCustomers;
    }
}


// Tests
[Fact]
public async Task TestPrimitivePropertyInDerivedTypeAsync()
{
    var queryUrl = "default/MyCustomers?$apply=groupby((Address/Microsoft.AspNetCore.OData.E2E.Tests.DollarApply.MyHomeAddress/HomeNo))";
    var request = new HttpRequestMessage(HttpMethod.Get, queryUrl);
    request.Headers.Accept.Add(MediaTypeWithQualityHeaderValue.Parse("application/json;odata.metadata=none"));
    var client = CreateClient(); // Initialize client

    var response = await client.SendAsync(request);

    // Assert
    Assert.Equal(HttpStatusCode.OK, response.StatusCode);
            
    var content = await response.Content.ReadAsStringAsync();
    // Response as expected: {"value":[{"Address":{"HomeNo":"#1"}},{"Address":{"HomeNo":"#2"}}]}
}

[Fact]
public async Task TestComplexPropertyInDerivedTypeAsync()
{
    var queryUrl = "default/MyCustomers?$apply=groupby((Address/Microsoft.AspNetCore.OData.E2E.Tests.DollarApply.MyHomeAddress/City/Name))";
    var request = new HttpRequestMessage(HttpMethod.Get, queryUrl);
    request.Headers.Accept.Add(MediaTypeWithQualityHeaderValue.Parse("application/json;odata.metadata=none"));
    var client = CreateClient(); // Initialize client

    var response = await client.SendAsync(request);

    // Assert
    Assert.Equal(HttpStatusCode.OK, response.StatusCode);

    var content = await response.Content.ReadAsStringAsync();
    // Response NOT as expected: {"value":[{"Address":{}},{"Address":{}}]}
}
```

clemvnt commented:
Thank you for the clarification.

It seems the issue arises from the `ODataResourceSerializer`, which is unable to locate the `City` property on the structured type `Microsoft.AspNetCore.OData.E2E.Tests.DollarApply.MyAddress` :
https://github.com/OData/AspNetCoreOData/blob/0ceeb086e4fd7a6396b4c6e0fd4f123283d55736/src/Microsoft.AspNetCore.OData/Formatter/Serialization/ODataResourceSerializer.cs#L334-L335

Looking at the `GroupByWrapper` instance, there doesn't appear to be any information indicating the EDM type of the value. I’m unsure how to infer that the type should be `MyHomeAddress` instead of `MyAddress` within the `ODataResourceSerializer` logic. Do you have any insights on how to address this?

_For context, in my specific use case, I don’t encounter this issue because I bypass the OData formatter. Instead, I manually apply OData clauses to `IQueryable` objects._

gathogojr commented:
> Thank you for the clarification.
> 
> It seems the issue arises from the `ODataResourceSerializer`, which is unable to locate the `City` property on the structured type `Microsoft.AspNetCore.OData.E2E.Tests.DollarApply.MyAddress` :
> 
> https://github.com/OData/AspNetCoreOData/blob/0ceeb086e4fd7a6396b4c6e0fd4f123283d55736/src/Microsoft.AspNetCore.OData/Formatter/Serialization/ODataResourceSerializer.cs#L334-L335
> 
> Looking at the `GroupByWrapper` instance, there doesn't appear to be any information indicating the EDM type of the value. I’m unsure how to infer that the type should be `MyHomeAddress` instead of `MyAddress` within the `ODataResourceSerializer` logic. Do you have any insights on how to address this?
> 
> _For context, in my specific use case, I don’t encounter this issue because I bypass the OData formatter. Instead, I manually apply OData clauses to `IQueryable` objects._


@clemvnt I'll need to do some digging to figure out how we can remedy it. At least we now know about it. If it's okay with you, we could merge your fix since it addresses a gap, then open a new ticket to track this scenario.

gathogojr commented:
Thinking about it some more...

> It seems the issue arises from the `ODataResourceSerializer`, which is unable to locate the `City` property on the structured type `Microsoft.AspNetCore.OData.E2E.Tests.DollarApply.MyAddress`
> I’m unsure how to infer that the type should be `MyHomeAddress` instead of `MyAddress` within the `ODataResourceSerializer` logic

Somehow it works for the primitive `HomeNo` property declared on `MyHomeAddress`... We'll need to figure out why that works and why it doesn't work if the property is complex.

clemvnt commented:
> If it's okay with you, we could merge your fix since it addresses a gap, then open a new ticket to track this scenario.

Sure, I’d be happy to create a new issue for this case and include all the information I have gathered about it.

> Somehow it works for the primitive `HomeNo` property declared on `MyHomeAddress`... We'll need to figure out why that works and why it doesn't work if the property is complex.

The EDM property isn’t found even for the primitive `HomeNo` property. However, it works because the following condition is satisfied:
https://github.com/OData/AspNetCoreOData/blob/0ceeb086e4fd7a6396b4c6e0fd4f123283d55736/src/Microsoft.AspNetCore.OData/Formatter/Serialization/ODataResourceSerializer.cs#L368-L372

For the `City` property, we don’t meet the same condition since it’s a navigation property.

WanjohiSammy commented:
...
> Sure, I’d be happy to create a new issue for this case and include all the information I have gathered about it.
> 
@clemvnt if you have already created that issue, kindly link with thie PR or somehow update it so that we are able to relate the 2

gathogojr commented:
> ...
> 
> > Sure, I’d be happy to create a new issue for this case and include all the information I have gathered about it.
> 
> @clemvnt if you have already created that issue, kindly link with thie PR or somehow update it so that we are able to relate the 2

I have created the issue https://github.com/OData/AspNetCoreOData/issues/1408
